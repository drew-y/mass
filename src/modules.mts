import { ParsedFiles } from "./lib/parse-directory.mjs";
import { List } from "./syntax-objects/list.mjs";
import { VoidModule } from "./syntax-objects/module.mjs";

export const resolveFileModules = (opts: {
  /** Path to the std lib directory */
  stdPath: string;
  /** Path to the user source code */
  srcPath: string;
  files: ParsedFiles;
}): VoidModule => {
  const { stdPath, srcPath, files } = opts;

  const rootModule = new VoidModule({ name: "root" });

  for (const [filePath, file] of Object.entries(files)) {
    const resolvedPath = filePathToModulePath(filePath, srcPath, stdPath);
    const parsedPath = resolvedPath.split("/").filter(Boolean);
    registerModule({ path: parsedPath, parentModule: rootModule, ast: file });
  }

  return rootModule;
};

const registerModule = ({
  path,
  parentModule,
  ast,
}: {
  path: string[];
  parentModule: VoidModule;
  ast: List;
}): VoidModule | undefined => {
  const [name, ...rest] = path;

  if (!name) return;

  const existingModule = parentModule.resolveChildEntity(name);

  if (existingModule && !existingModule.isModule()) {
    throw new Error(
      `Cannot register module ${name} because it is already registered as ${existingModule.syntaxType}`
    );
  }

  if (!existingModule && name === "index") {
    parentModule.push(...ast.value);
    return;
  }

  const module =
    existingModule ??
    new VoidModule({
      ...(!rest.length ? { ...ast.metadata, value: ast.value } : {}),
      name,
    });
  module.isExported = true;

  if (!existingModule) parentModule.push(module);

  if (!rest.length) return;

  return registerModule({ path: rest, parentModule: module, ast });
};

const filePathToModulePath = (
  filePath: string,
  srcPath: string,
  stdPath: string
) => {
  let finalPath = filePath.startsWith(stdPath)
    ? filePath.replace(stdPath, "std")
    : filePath;

  finalPath = finalPath.startsWith(srcPath)
    ? finalPath.replace(srcPath, "src")
    : finalPath;

  finalPath = finalPath.replace(".void", "");

  return finalPath;
};
