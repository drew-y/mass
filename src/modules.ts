import { ParsedFiles } from "./lib/parse-directory.js";
import { List } from "./syntax-objects/list.js";
import { VoidModule } from "./syntax-objects/module.js";

export const resolveFileModules = (opts: {
  /** Path to the std lib directory */
  stdPath: string;
  /** Path to the user source code root folder */
  srcPath?: string;
  /** Path to the entry index file, which determines what is exported from the package */
  indexPath: string;
  files: ParsedFiles;
}): VoidModule => {
  const { stdPath, srcPath, files } = opts;

  const rootModule = new VoidModule({ name: "root" });

  for (const [filePath, file] of Object.entries(files)) {
    const resolvedPath = filePathToModulePath(
      filePath,
      srcPath ?? opts.indexPath,
      stdPath
    );

    const parsedPath = resolvedPath.split("/").filter(Boolean);

    registerModule({
      path: parsedPath,
      parentModule: rootModule,
      ast: file.slice(1), // Skip the first element (ast)
      isIndex: filePath === opts.indexPath,
    });
  }

  return rootModule;
};

const registerModule = ({
  path,
  parentModule,
  ast,
  isIndex,
}: {
  path: string[];
  parentModule: VoidModule;
  ast: List;
  isIndex?: boolean;
}): VoidModule | undefined => {
  const [name, ...rest] = path;

  if (!name) return;

  const existingModule = parentModule.resolveChildEntity(name);

  if (existingModule && !existingModule.isModule()) {
    throw new Error(
      `Cannot register module ${name} because it is already registered as ${existingModule.syntaxType}`
    );
  }

  if (!existingModule && name === "index") {
    parentModule.push(...ast.toArray());
    return;
  }

  const module =
    existingModule ??
    new VoidModule({
      ...(!rest.length ? { ...ast.metadata, value: ast.toArray() } : {}),
      name,
      isIndex,
    });
  module.isExported = true;

  if (!existingModule) parentModule.push(module);

  if (!rest.length) return;

  return registerModule({ path: rest, parentModule: module, ast });
};

const filePathToModulePath = (
  filePath: string,
  srcPath: string,
  stdPath: string
) => {
  let finalPath = filePath.startsWith(stdPath)
    ? filePath.replace(stdPath, "std")
    : filePath;

  finalPath = finalPath.startsWith(srcPath)
    ? finalPath.replace(srcPath, "src")
    : finalPath;

  finalPath = finalPath.replace(".void", "");

  return finalPath;
};
