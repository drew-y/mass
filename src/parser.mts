import { ModuleInfo } from "./lib/module-info.mjs";
import { Expr, Identifier, List, Whitespace } from "./lib/syntax/index.mjs";
import { Token } from "./lib/token.mjs";
import { File } from "./lib/file.mjs";
import { getReaderMacroForToken } from "./reader-macros/index.mjs";

export interface ParseOpts {
  nested?: boolean;
  terminator?: string;
  module: ModuleInfo;
  parent?: Expr;
}

export function parse(file: File, opts: ParseOpts): List {
  const list = new List({
    location: {
      startIndex: file.position,
      endIndex: 0,
      line: file.line,
      column: file.column,
      filePath: file.filePath,
    },
    parent: opts.parent,
  });

  while (file.hasCharacters) {
    const token = lexer(file);

    const readerMacro = getReaderMacroForToken(token);

    if (readerMacro) {
      const result = readerMacro(file, {
        token,
        module: opts.module,
        reader: (file, terminator, parent) =>
          parse(file, {
            nested: true,
            terminator,
            module: opts.module,
            parent: parent ?? opts.parent,
          }),
      });
      if (typeof result !== "undefined") list.push(result);
      continue;
    }

    if (token.is("(")) {
      list.push(parse(file, { nested: true, module: opts.module }));
      continue;
    }

    if (token.is(")") || token.is(opts.terminator)) {
      if (opts.nested) break;
      continue;
    }

    if (token.isWhitespace) {
      list.push(
        new Whitespace({
          value: token.value,
          location: token.location,
        })
      );
    }

    list.push(
      new Identifier({
        value: token.value,
        location: token.location,
      })
    );
  }

  list.location!.endIndex = file.position;
  return list;
}

const lexer = (file: File): Token => {
  const token = new Token({
    line: file.line,
    column: file.column,
    startIndex: file.position,
    endIndex: 0,
    filePath: file.filePath,
  });

  while (file.hasCharacters) {
    const char = file.next;

    // Ignore commas for now. They make a nice visual separator
    if (char === ",") {
      file.consume();
      continue;
    }

    // Handle real numbers
    if (char === "." && token.isNumber) {
      token.addChar(file.consume());
      continue;
    }

    const isTerminator = /[\{\[\(\}\]\)\s\.\;\:\'\"]/.test(char);

    if (isTerminator && (token.first === "#" || !token.hasChars)) {
      token.addChar(file.consume());
      break;
    }

    if (isTerminator) {
      break;
    }

    token.addChar(file.consume());
  }

  token.location.endIndex = file.position;
  return token;
};
