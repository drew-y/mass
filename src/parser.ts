import { readerMacros } from "./reader-macros";

export type AST = Expr[];
export type Expr = string | AST;

export interface ParseOpts {
  nested?: boolean;
  insertToken?: string;
  terminator?: string;
}

export function parse(dream: string[], opts: ParseOpts = {}): AST {
  const ast: AST = [];
  let token = "";

  const runReaderMacro = (tag: string) => {
    ast.push(
      readerMacros.get(tag)!(dream, (dream, terminator) =>
        parse(dream, { nested: true, terminator })
      )
    );
  };

  const pushCurrentToken = () => {
    if (token[0] === "#" && readerMacros.has(token)) {
      runReaderMacro(token);
      token = "";
      return;
    }

    if (token) ast.push(token);
    token = "";
  };

  if (opts.insertToken) {
    ast.push(opts.insertToken);
  }

  while (dream.length) {
    const char = dream.shift();

    if (char === " " || char === "\t" || char === "\n") {
      pushCurrentToken();
      ast.push(char);
      continue;
    }

    if (token[0] === "#") {
      token += char;
      if (["(", "[", "{"].includes(char!)) pushCurrentToken();
      continue;
    }

    if (readerMacros.has(char ?? "")) {
      pushCurrentToken();
      runReaderMacro(char!);
      continue;
    }

    if (char === "(" && token) {
      ast.push(parse(dream, { insertToken: token, nested: true }));
      token = "";
      continue;
    }

    if (char === "(") {
      ast.push(parse(dream, { nested: true }));
      continue;
    }

    if (char === ")" || char === opts.terminator || token === opts.terminator) {
      pushCurrentToken();
      if (opts.nested) break;
      continue;
    }

    token += char;
  }

  return ast;
}
