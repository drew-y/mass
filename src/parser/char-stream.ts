import { SourceLocation } from "../syntax-objects/syntax.js";

export class CharStream {
  readonly filePath: string;
  readonly originalSize: number;
  readonly value: string[];
  readonly location = {
    index: 0,
    line: 1,
    column: 0,
  };

  constructor(contents: string, filePath: string) {
    this.value = contents.split("").reverse();
    this.originalSize = this.value.length;
    this.filePath = filePath;
  }

  /** Current index the file is on */
  get position() {
    return this.location.index;
  }

  get line() {
    return this.location.line;
  }

  get column() {
    return this.location.column;
  }

  get hasCharacters() {
    return !!this.value.length;
  }

  get next() {
    return this.value[this.value.length - 1];
  }

  currentSourceLocation() {
    return new SourceLocation({
      startIndex: this.position,
      endIndex: this.originalSize - this.value.length,
      line: this.line,
      column: this.column,
      filePath: this.filePath,
    });
  }

  at(index: number): string | undefined {
    return this.value.at(-index - 1);
  }

  /** Returns the next character and removes it from the queue */
  consumeChar(): string {
    const char = this.value.pop();
    if (!char) {
      throw new Error("Out of characters");
    }

    this.location.index += 1;
    this.location.column += 1;
    if (char === "\n") {
      this.location.line += 1;
      this.location.column = 0;
    }

    return char;
  }
}
