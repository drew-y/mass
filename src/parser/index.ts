export { parse } from "./parser.js";
export * from "./utils/index.js";
