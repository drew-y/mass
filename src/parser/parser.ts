import { Token, tokenize } from "../lexer";
import { Instruction, VariableDeclaration, TypeArgument, MethodDeclaration, ParameterDeclaration, ReturnStatement, Assignment, EnumDeclaration, EnumVariant, MatchCase } from "./definitions";
import { isInTuple } from "../helpers";

export function parse(code: string): Instruction[] {
    const tokens = tokenize(code);
    return parseTokens(tokens);
}

function parseTokens(tokens: Token[]): Instruction[] {
    const ast: Instruction[] = [];

    while (tokens.length > 0) {
        const next = tokens[0];
        if (next.type === "}") {
            tokens.shift();
            break;
        }

        if (next.type === "\n") {
            tokens.shift();
            continue;
        }

        ast.push(parseStatement(tokens));
    }

    return ast;
}

function parseStatement(tokens: Token[]): Instruction {
    while (tokens.length > 0) {
        let token = tokens[0];
        const next = tokens[1];

        // Ignore newlines and semicolons
        if (isInTuple(token.type, ["\n", ";"] as const)) {
            tokens.shift();
            continue;
        }

        if (token.type === "keyword") {
            return parseKeywordStatement(tokens);
        }

        if (token.type === "identifier" && next && next.type === "=") {
            return parseAssignment(tokens);
        }

        return parseExpression(tokens);
    }

    throw new Error("Invalid statement");
}

function parseKeywordStatement(tokens: Token[]): Instruction {
    let token = tokens[0];

    if (token.value === "return") {
        return parseReturnStatement(tokens);
    }

    if (["if", "while", "for", "match"].includes(token.value)) {
        return parseExpression(tokens);
    }

    if (token.value === "continue") {
        return { kind: "continue-statement" };
    }

    if (token.value === "break") {
        return { kind: "break-statement" };
    }

    const flags: string[] = [];
    while (token.type === "keyword") {
        flags.push(tokens.shift()!.value);
        token = tokens[0];
    }

    if (flags.some(val => val === "let" || val === "var")) {
        return parseVariableDeclaration(tokens, flags);
    }

    if (flags.includes("def")) {
        return parseMethodDeclaration(tokens, flags);
    }

    if (flags.includes("enum")) {
        return parseEnumDeclaration(tokens, flags);
    }

    const keywordStr = flags.reduce((p, c) => `${p} ${c}`, "");
    throw new Error(`Expected statement after keyword(s):${keywordStr}`);
}

function parseAssignment(tokens: Token[]): Assignment {
    const identifier = tokens.shift();
    if (!identifier || identifier.type !== "identifier") {
        throw new Error(`Unexpected identifier token in assignment`);
    }

    const equals = tokens.shift();
    if (!equals || equals.type !== "=") {
        throw new Error("Expected = token in assignment statement");
    }

    return {
        kind: "assignment",
        identifier: identifier.value,
        expression: parseExpression(tokens)
    }
}

function parseReturnStatement(tokens: Token[]): ReturnStatement {
    const returnToken = tokens.shift();
    if (!returnToken || returnToken.value !== "return") {
        throw new Error("Expected return token");
    }

    return {
        kind: "return-statement",
        expression: parseExpression(tokens)
    }
}

/** Parse a method, beginning after def */
function parseMethodDeclaration(tokens: Token[], flags: string[]): MethodDeclaration {
    const identifierToken = tokens.shift();
    if (!identifierToken || identifierToken.type !== "identifier") {
        throw new Error("Expected identifier after method declaration");
    }

    const identifier = identifierToken.value;
    const parameters = parseMethodParameters(tokens);

    let returnType: TypeArgument | undefined;
    if (tokens[0].type === "->") {
        tokens.shift();
        returnType = parseTypeArgument(tokens);
    }

    if (tokens[0].type !== "{") {
        throw new Error(`Unexpected token in method declaration: ${tokens[0].type}`);
    }
    tokens.shift();

    const body = parseTokens(tokens);

    return {
        kind: "method-declaration",
        identifier,
        parameters,
        returnType,
        body,
        typeParameters: [], // TODO
        flags
    }
}

function parseMethodParameters(tokens: Token[]): ParameterDeclaration[] {
    const params: ParameterDeclaration[] = [];

    const openingBracket = tokens.shift();
    if (!openingBracket || !isInTuple(openingBracket.type, <const>["("])) {
        throw new Error("Method definition missing parameters");
    }

    // In the future, we will support "]" as well
    const closeBracket = <const>")";

    let token = tokens[0];
    while (token && token.type !== closeBracket) {
        if (token.type === "identifier") {
            params.push(parseParameter(tokens));
            token = tokens[0];
            continue;
        }

        if (token.type === ",") {
            tokens.shift();
            token = tokens[0];
            continue;
        }

        throw new Error(`Invalid token in parameters: ${token.type}`);
    }

    // Remove the closeBracket
    tokens.shift();

    return params;
}

function parseParameter(tokens: Token[]): ParameterDeclaration {
    const flags: string[] = [];

    const identifierToken = tokens.shift();
    if (!identifierToken || identifierToken.type !== "identifier") {
        throw new Error("Invalid parameter definition");
    }

    const identifier = identifierToken.value;

    const separator = tokens.shift();
    if (!separator || !isInTuple(separator.value, <const>[":", "="])) {
        throw new Error("Unexpected token in parameter definition");
    }

    if (separator.value === "=") {
        const initializer = parseExpression(tokens);
        return {
            kind: "parameter-declaration",
            identifier,
            initializer,
            flags
        }
    }

    let token = tokens[0];
    while (token && isInTuple(token.value, <const>["mut", "ref"])) {
        flags.push(token.value);
        tokens.shift();
        token = tokens[0];
    }

    const type = parseTypeArgument(tokens);

    return {
        kind: "parameter-declaration",
        identifier,
        type,
        flags
    }
}

function parseVariableDeclaration(tokens: Token[], flags: string[]): VariableDeclaration {
    const identifiers: string[] = [];
    let type: TypeArgument | undefined = undefined;
    let initializer: Instruction | undefined;

    while (tokens[0] && isInTuple(tokens[0].type, <const>["identifier", ","])) {
        const token = tokens.shift()!;

        if (token.type === ",") {
            tokens.shift();
            continue;
        }

        identifiers.push(token.value);
    }

    if (tokens[0].type === ":") {
        tokens.shift();
        type = parseTypeArgument(tokens);
    }

    if (tokens[0].value === "=") {
        tokens.shift();
        initializer = parseExpression(tokens);
    }

    return {
        kind: "variable-declaration",
        identifiers, flags, type, initializer
    };
}

function parseTypeArgument(tokens: Token[]): TypeArgument {
    const token = tokens.shift()!;
    // For now we assume a simple type as an identifier.
    return {
        kind: "type-argument",
        identifier: token.value,
        flags: []
    };
}

function parseExpression(tokens: Token[], terminator?: Token): Instruction {
    const output: Instruction[] = [];
    const operator: Token[] = [];

    // Since we don't use ; to terminate an expression, we can tell the expression
    // Is done if we get two non-operator tokens in a row (-newlines).
    // This is my ugly temp solution. Looking for a way to make it cleaner.
    let expectOperatorToContinue = false;
    while (tokens.length > 0) {
        const token = tokens[0];

        // Ignore new lines for now
        if (token.type === "\n") {
            tokens.shift();
            continue;
        }

        // Consumed terminator
        // TODO: !!! Important !!! Make this non consuming
        if (terminator && token.type === terminator.type) {
            tokens.shift();
            break;
        }

        if (token.type !== "operator") {
            if (expectOperatorToContinue) break;
            expectOperatorToContinue = true;
        } else {
            if (!expectOperatorToContinue) {
                throw new Error(`Unexpected operator: ${token.value}`);
            }

            expectOperatorToContinue = false;
        }

        if (token.type === "int") {
            output.push({ kind: "i32-literal", value: token.value });
            tokens.shift();
            continue;
        }

        if (token.type === "float") {
            output.push({ kind: "f32-literal", value: token.value });
            tokens.shift();
            continue;
        }

        if (token.type === "string") {
            output.push({ kind: "string-literal", value: token.value });
            tokens.shift();
            continue;
        }

        if (token.type === "boolean") {
            output.push({ kind: "bool-literal", value: token.value === "true" });
            tokens.shift();
            continue;
        }

        if (token.type === "keyword") {
            if (token.value === "if") {
                tokens.shift();
                const condition = parseExpression(tokens, { type: "{", value: "{" });
                const body = parseTokens(tokens);
                output.push({ kind: "if-expression", condition, body });
                continue;
            }

            if (token.value === "while") {
                tokens.shift()
                const condition = parseExpression(tokens, { type: "{", value: "{" });
                const body = parseTokens(tokens);
                output.push({ kind: "while-statement", condition, body });
                continue;
            }

            if (token.value === "match") {
                tokens.shift();
                const expression = parseExpression(tokens, { type: "{", value: "{" });
                const cases = parseMatchCases(tokens);
                output.push({ kind: "match-expression", expression, cases, flags: [] });
                continue;
            }

            throw new Error(`Invalid keyword in expression: ${token.value}`);
        }

        if (token.type === "identifier") {
            // Handle possible function / method call
            const next = tokens[1];
            if (next && isInTuple(next.type, ["(", "["])) {
                // Remove identifier token
                tokens.shift();

                output.push({
                    kind: "method-or-function-call",
                    identifier: token.value,
                    arguments: parseArguments(tokens), // T
                });
                continue;
            }

            output.push({ kind: "identifier", value: token.value });
            tokens.shift();
            continue;
        }

        if (token.type === "operator") {
            while (operator.length > 0) {
                const op = operator[operator.length - 1];
                if (getOperatorPrecedence(op.value) >= getOperatorPrecedence(token.value)) {
                    output.push({
                        kind: "method-or-function-call",
                        identifier: operator.pop()!.value,
                        arguments: [output.pop()!, output.pop()!]
                    });
                    continue;
                }
                break;
            }

            operator.push(tokens.shift()!);
            continue;
        }

        if (token.type === "(") {
            tokens.shift();
            output.push(parseExpression(tokens));
            tokens.shift();
            continue;
        }

        // Non-consumed terminators
        if (isInTuple(token.type, ["}", ")", ","])) break;

        throw new Error(`Unexpected token: ${token.type} ${token.value}`);
    }

    // Infix parsing
    while (operator.length > 0) {
        output.push({
            kind: "method-or-function-call",
            identifier: operator.pop()!.value,
            arguments: [output.pop()!, output.pop()!]
        });
    }

    return output[0] as Instruction;
}

function getOperatorPrecedence(operator: string): number {
    const precedences: Record<string, number> = {
        "and": 1,
        "or": 1,
        "xor": 1,
        "==": 2,
        "<": 2,
        ">": 2,
        ">=": 2,
        "<=": 2,
        "<>": 2,
        "?": 2,
        "+": 2,
        "-": 2,
        "*": 3,
        "/": 3,
        "^": 4,
        ".": 5,
        "=": 0,
        "=>": 0
    }
    return precedences[operator];
}

function parseArguments(tokens: Token[]): Instruction[] {
    const args: Instruction[] = [];

    // For now we just get rid of the opening brace. We also only handle (
    const openingBracket = tokens.shift();
    if (!openingBracket || openingBracket.type !== "(") {
        throw new Error("Expected opening bracket in argument expression");
    }

    while (tokens.length > 0) {
        let token = tokens[0];
        if (token.type === ",") {
            tokens.shift();
        }

        if (token.type === ")") {
            tokens.shift();
            break;
        }

        args.push(parseExpression(tokens))
        token = tokens[0];
    }

    return args;
}

/** Parse an enum, beginning after enum */
function parseEnumDeclaration(tokens: Token[], flags: string[]): EnumDeclaration {
    const identifierToken = tokens.shift();
    if (!identifierToken || identifierToken.type !== "identifier") {
        throw new Error("Expected identifier after enum declaration");
    }

    const identifier = identifierToken.value;

    if (tokens[0].type !== "{") {
        throw new Error(`Unexpected token in method declaration: ${tokens[0].type}`);
    }
    tokens.shift();

    const variants = parseEnumVariants(tokens, identifier);

    return {
        kind: "enum-declaration",
        identifier,
        flags,
        variants,
        typeParameters: []
    }
}

function parseEnumVariants(tokens: Token[], parentEnum: string): EnumVariant[] {
    const variants: EnumVariant[] = [];

    while (tokens[0]) {
        const token = tokens[0];
        if (token.type === "identifier") {
            variants.push({
                kind: "enum-variant",
                identifier: token.value,
                parentEnum,
                flags: []
            });
            tokens.shift();
            continue;
        }

        if (token.type === "," || token.type === "\n") {
            tokens.shift();
            continue;
        }

        if (token.type === "}") {
            tokens.shift();
            break;
        }

        throw new Error(`Unexpected token in enum: ${token.value}`);
    }

    return variants;
}

function parseMatchCases(tokens: Token[]): MatchCase[] {
    const variants: MatchCase[] = [];

    while (tokens[0]) {
        const token = tokens[0];

        if (token.type === "}") {
            tokens.shift();
            break;
        }

        if (token.type === "\n") {
            tokens.shift();
            continue;
        }

        const matchCase = parseExpression(tokens, { type: "=>", value: "=>" });
        const expression = parseExpression(tokens, { type: ",", value: "," });

        variants.push({
            kind: "match-case",
            case: matchCase,
            expression
        });
    }

    return variants;
}
