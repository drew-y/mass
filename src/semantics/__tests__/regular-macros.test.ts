import { parse } from "../../parser/parser.js";
import path from "node:path";
import { registerModules } from "../modules.js";
import { expandRegularMacros } from "../regular-macros.js";
import { regularMacrosVoidFile } from "./fixtures/regular-macros-void-file.js";
import { test } from "vitest";

test("regular macro expansion", async (t) => {
  const parserOutput = parse(regularMacrosVoidFile);
  const files = { test: parserOutput };
  const resolvedModules = registerModules({
    files,
    srcPath: path.dirname("test"),
    indexPath: "test.void",
  });
  const result = expandRegularMacros(resolvedModules);
  t.expect(result).toMatchSnapshot();
});
