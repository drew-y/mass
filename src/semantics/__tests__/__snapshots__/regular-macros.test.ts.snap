// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`regular macro expansion 1`] = `
[
  "module",
  "root",
  [
    "exports",
    [],
  ],
  [
    [
      "module",
      "test",
      [
        "exports",
        [],
      ],
      [
        [
          "use",
          [
            "::",
            "root",
            "all",
          ],
        ],
        [
          "regular-macro",
          "\`#744",
          [
            "parameters",
          ],
          [
            "block",
            [
              "block",
              [
                "quote",
                "quote",
                [
                  "$@",
                  "body",
                ],
              ],
            ],
          ],
        ],
        [
          "regular-macro",
          "let#781",
          [
            "parameters",
          ],
          [
            "block",
            [
              "block",
              [
                "define",
                "equals_expr",
                [
                  "extract",
                  "body",
                  0,
                ],
              ],
              [
                "quote",
                "define",
                [
                  "$",
                  [
                    "extract",
                    "equals_expr",
                    1,
                  ],
                ],
                [
                  "$",
                  [
                    "extract",
                    "equals_expr",
                    2,
                  ],
                ],
              ],
            ],
          ],
        ],
        [
          "define-macro-variable",
          "extract_parameters",
          [
            "reserved-for-type",
          ],
          [
            "is-mutable",
            false,
          ],
        ],
        [
          "regular-macro",
          "fn#1260",
          [
            "parameters",
          ],
          [
            "block",
            [
              "block",
              [
                "define",
                "definitions",
                [
                  "extract",
                  "body",
                  0,
                ],
              ],
              [
                "define",
                "identifier",
                [
                  "extract",
                  "definitions",
                  0,
                ],
              ],
              [
                "define",
                "params",
                [
                  "extract_parameters",
                  "definitions",
                ],
              ],
              [
                "define",
                "type_arrow_index",
                [
                  "block",
                  [
                    "if",
                    [
                      "==",
                      [
                        "extract",
                        "body",
                        1,
                      ],
                      [
                        "string",
                        "->",
                      ],
                    ],
                    [
                      ":",
                      "then",
                      [
                        "block",
                        1,
                      ],
                    ],
                    [
                      ":",
                      "else",
                      [
                        "block",
                        [
                          "if",
                          [
                            "==",
                            [
                              "extract",
                              "body",
                              2,
                            ],
                            [
                              "string",
                              "->",
                            ],
                          ],
                          [
                            ":",
                            "then",
                            2,
                          ],
                          [
                            ":",
                            "else",
                            -1,
                          ],
                        ],
                      ],
                    ],
                  ],
                ],
              ],
              [
                "define",
                "return_type",
                [
                  "block",
                  [
                    "if",
                    [
                      ">",
                      "type_arrow_index",
                      -1,
                    ],
                    [
                      ":",
                      "then",
                      [
                        "block",
                        [
                          "slice",
                          "body",
                          [
                            "+",
                            "type_arrow_index",
                            1,
                          ],
                          [
                            "+",
                            "type_arrow_index",
                            2,
                          ],
                        ],
                      ],
                    ],
                    [
                      ":",
                      "else",
                      [
                        "quote",
                      ],
                    ],
                  ],
                ],
              ],
              [
                "define",
                "expressions",
                [
                  "block",
                  [
                    "if",
                    [
                      ">",
                      "type_arrow_index",
                      -1,
                    ],
                    [
                      ":",
                      "then",
                      [
                        "block",
                        [
                          "slice",
                          "body",
                          [
                            "+",
                            "type_arrow_index",
                            2,
                          ],
                        ],
                      ],
                    ],
                    [
                      ":",
                      "else",
                      [
                        "slice",
                        "body",
                        1,
                      ],
                    ],
                  ],
                ],
              ],
              [
                "quote",
                "define_function",
                [
                  "$",
                  "identifier",
                ],
                [
                  "$",
                  "params",
                ],
                [
                  "return_type",
                  [
                    "$@",
                    "return_type",
                  ],
                ],
                [
                  "$",
                  [
                    "concat",
                    [
                      "quote",
                      "block",
                    ],
                    "expressions",
                  ],
                ],
              ],
            ],
          ],
        ],
        [
          "define_function",
          "fib",
          [
            "parameters",
            [
              ":",
              "n",
              "i32",
            ],
          ],
          [
            "return_type",
            "i32",
          ],
          [
            "block",
            [
              "block",
              [
                "define",
                "base",
                1,
              ],
              [
                "if",
                [
                  "<=",
                  "n",
                  "base",
                ],
                [
                  ":",
                  "then",
                  [
                    "block",
                    "n",
                  ],
                ],
                [
                  ":",
                  "else",
                  [
                    "block",
                    [
                      "+",
                      [
                        "fib",
                        [
                          "-",
                          "n",
                          1,
                        ],
                      ],
                      [
                        "fib",
                        [
                          "-",
                          "n",
                          2,
                        ],
                      ],
                    ],
                  ],
                ],
              ],
            ],
          ],
        ],
      ],
    ],
  ],
]
`;
